baselines/fedvssl/README.md

```diff
@@ -0,0 +1,171 @@
+For UCF-101 dataset, one can simply run the `dataset_preparation.py` file to download and generate the iid splits for UCF-101 datasets.
+
+| Dataset | #classes | #partitions | partitioning method |  partition settings  |
+|:--------|:--------:|:-----------:| :---: |:--------------------:|
+| UCF101  |   101    |     10      | randomly partitioned |       uniform        |
+| Kinectics-400    |   400    |     100     | randomly partitioned | 8 classes per client |
+
+**Training Hyperparameters:** The following table shows the main hyperparameters for this baseline with their default value (i.e. the value used if you run python main.py directly)
+
+| Description        |            Default Value            |
+|:-------------------|:-----------------------------------:|
+| total clients      |                 2                  |
+| clients per round  |                 2                  | 
+| number of rounds	  |                 20                  | 
+| client resources	  | {'num_cpus': 2.0, 'num_gpus': 1.0 } | 
+| optimizer	  |                 SGD                  | 
+| alpha coefficient	 |                 0.9                 | 
+| beta coefficient	  |                  1                  | 
+
+## Environment Setup
+Please make sure you have installed CUDA 11.0 on your machine 
+(see [NVIDIA CUDA Toolkit website](https://developer.nvidia.com/cuda-11.0-download-archive?target_os=Linux)).
+To construct the Python environment follow these steps:
+
+```bash
+# install the base Poetry environment
+poetry install
+
+# activate the environment
+poetry shell
+
+# install PyTorch with GPU support.
+pip install torch==1.7.1+cu110 torchvision==0.8.2+cu110 torchaudio==0.7.2 -f https://download.pytorch.org/whl/torch_stable.html
+
+# install mmcv package
+pip install mmcv-full==1.2.4 -f https://download.openmmlab.com/mmcv/dist/cu110/torch1.7.1/index.html
+pip install numpy==1.21.0
+pip install tensorboard
+pip install mmengine==0.9.0
+```
+
+## Running the Experiments
+To run FedVSSL with UCF-101 baseline, please ensure you have activated your Poetry environment (execute `poetry shell` from this directory). Then, download the `CtP` repo, download the datasets and preprocess it:
+
+```bash
+# clone CtP repo
+git clone https://github.com/yan-gao-GY/CtP.git fedvssl/CtP
+
+sudo apt install unrar unzip
+
+cd fedvssl
+python dataset_preparation.py
+
+cd ..
+```
+
+
+Finally, we can launch the training:
+
+```bash
+# run federated SSL training with FedVSSL
+python -m fedvssl.main pre_training=true # this will run using the default settings.
+
+# you can override settings directly from the command line
+python -m fedvssl.main pre_training=true mix_coeff=1 rounds=100 # will set hyper-parameter alpha to 1 and the number of rounds to 100
+
+# run downstream fine-tuning with pre-trained SSL model
+python -m fedvssl.main pre_training=false pretrained_model_path=<CHECKPOINT>.npz # this will run using the default settings.
```

Comment:
I'm unable to run this. Resulting in errors of the type:
`NCCL error in: /pytorch/torch/lib/c10d/ProcessGroupNCCL.cpp:784, invalid usage, NCCL version 2.7.8`

Could you give details on the exact CUDA version you used and Ubuntu version? Flower baselines expect Ubuntu 22.04 (for which only CUDA 11.7+ is available irrc).

Reply:
@jafermarq We haven't tested that on Ubuntu 22.04. We tested it with the following configuration:
Ubuntu -- > VERSION="16.04.6 LTS (Xenial Xerus)
CUDA Version: 11.4

I will try with CUDA 11.7 to see if can reproduce the above error.

Reply:
> I'm unable to run this. Resulting in errors of the type: `NCCL error in: /pytorch/torch/lib/c10d/ProcessGroupNCCL.cpp:784, invalid usage, NCCL version 2.7.8`
> 
> Could you give details on the exact CUDA version you used and Ubuntu version? Flower baselines expect Ubuntu 22.04 (for which only CUDA 11.7+ is available irrc).

@jafermarq Did you get the above error during the FL-pertaining or fine-tuning stage?

Reply:
@yasar-rehman only for the fine-tuning. I was able to run the pre-training with command:
```bash
python -m fedvssl.main pre_training=true
```

Reply:
@jafermarq @yan-gao-GY OK so after many attempts,  finally I was able to run both the pertaining and fine-tuning. 

@yan-gao-GY! I believe these updated versions of mmcv and mmengine will solve your side's "ROI_align" issue. 

CUDA Version: 12.0
Linux: 20.04.6 


Please install the following packages and remove the old ones. 

1.  pip install mmcv-full==1.7.1 -f https://download.openmmlab.com/mmcv/dist/cu117/torch1.13/index.html  (**Found this by chance**)
2. pip install torch==1.13.0+cu117 torchvision==0.14.0+cu117 torchaudio==0.13.0 --extra-index-url https://download.pytorch.org/whl/cu117  
3.  ypaf==0.40.1 (with version 0.40.2 I was getting the error: TypeError: FormatCode() got an unexpected keyword argument 'verify')
4. tensorboard without specifying any version
5. mmengine==0.10.0 

The mmcv-full==1.7.1, to the best of my knowledge, is the final latest version that can be used with the CtP repository in its current form. The new versions of the mmcv >=2.0.0 cannot be used with the current CtP repository because most of the functionality in mmengine > 0.10.0 has either been modified or removed. 

Let me know if you encounter any errors.

Reply:
amazing! this seems to be working. I had to change the number of processes to `1` (instead of `4`) in the main (because i only have a single GPU. I'll look into this more carefully and add another review later today.